jf_agent/git/github_gql_client.py

```diff
@@ -0,0 +1,466 @@
+            organization(login: \"{login}\") {{
+                id: databaseId
+                login
+                name
+                url
+            }}
+        }}
+        """
+        try:
+            return self._get_raw_result(query_body=query_body)['data']['organization']
+        except Exception as e:
+            log_and_print_request_error(
+                e, f'fetching source project {login}. ' f'Skipping...',
+            )
+            return None
+
+    def get_users(self, login: str) -> Generator[dict, None, None]:
+        query_body = f"""{{
+            organization(login: \"{login}\") {{
+                userQuery: membersWithRole(first: 100, after: %s) {{
+                    {self.GITHUB_GQL_PAGE_INFO_BLOCK}
+                    users: nodes {{
+                        {self.GITHUB_GQL_USER_FRAGMENT}
+                    }}
+                }}
+            }}
+        }}
+        """
+        for page in self._page_results(
+            query_body=query_body, path_to_page_info='data.organization.userQuery'
+        ):
+            for user in page['data']['organization']['userQuery']['users']:
+                yield user
+
+    def get_repos(
+        self, login: str, repo_filters: list[filter] = None
+    ) -> Generator[dict, None, None]:
+        query_body = f"""{{
+            organization(login: "{login}") {{
+                repoQuery: repositories(first: 50, after: %s) {{
+                    {self.GITHUB_GQL_PAGE_INFO_BLOCK}
+                    repos: nodes {{
+                        ... on Repository {{
+                            id: databaseId
+                            name
+                            fullName: nameWithOwner
+                            url
+                            isFork
+                            defaultBranch: defaultBranchRef {{
+                                    name
+                            }}
+                            # This should be broken out into a separate query if 
+                            # hasNextPage is True.
+                            branchQuery: refs(refPrefix:"refs/heads/", first: 50) {{
+                                {self.GITHUB_GQL_PAGE_INFO_BLOCK}
+                                branches: nodes {{
+                                    ... on Ref {{
+                                        name
+                                        target {{sha: oid}}
+                                    }}
+                                }}
+                            }}
+                        }}
+                    }}
+                }}
+            }}
+        }}
+        """
+        for page in self._page_results(
+            query_body=query_body, path_to_page_info='data.organization.repoQuery'
+        ):
+            for api_repo in page['data']['organization']['repoQuery']['repos']:
+                # Skip over excluded or ignore non-included
+                if not all(filt(api_repo) for filt in repo_filters):
```

Comment:
We might waste time here. Is there a way we can do this by generating a set of include and exclude and checking set membership here instead?

Reply:
Not sure exactly what you mean, I think that's what this is doing (I mostly copy-pasted this from the existing code)

I'm not actually super familiar with this all filter in python, but isn't it doing a simple 'is this string in the set' or not check for each repo name it finds?

```

        filters = []
        if self.config.git_include_repos:
            filters.append(
                lambda r: r['name'].lower()
                in set([r.lower() for r in self.config.git_include_repos])
            )
        if self.config.git_exclude_repos:
            filters.append(
                lambda r: r['name'].lower()
                not in set([r.lower() for r in self.config.git_exclude_repos])
            )
```

Reply:
Ah - I misunderstood the previous code. I thought it was creating an array of lambdas, but it only creates 2 and each is a set membership test. That makes sense. The code is slightly confusing because the `all` function makes it feel like we're iterating over a variety of filters when in reality it's only 2, but I don't think it's worth changing now.